qa/L0_lifecycle/lifecycle_test.py

```diff
@@ -2916,6 +2917,189 @@ def test_concurrent_load_unload(self):
+                    sequence_batching = {
+                        "direct": {},
+                        "max_sequence_idle_microseconds": sequence_timeout_secs
+                        * 1000000,
+                    }
+                    config = {
+                        "instance_group": instance_group,
+                        "max_batch_size": max_batch_size,
+                        "sequence_batching": sequence_batching,
+                        "input": input_config,
+                        "output": output_config,
+                    }
+                    print(
+                        f"~~~ Backend: [{backend}], Model: [{model}], Config: [{config}] ~~~"
+                    )
+                    # Load the model
+                    try:
+                        triton_client.load_model(model, config=json.dumps(config))
+                    except Exception as ex:
+                        self.assertTrue(False, "unexpected error {}".format(ex))
+
+                    # Make sure the model is loaded
+                    self.assertTrue(triton_client.is_server_live())
+                    self.assertTrue(triton_client.is_model_ready(model))
+                    print(
+                        "Model Repository Index after load:",
+                        triton_client.get_model_repository_index(),
+                    )
+
+                    # Test inference on each instance
+                    for i in range(1, num_instances + 1):
+                        try:
+                            triton_client.infer(
+                                model, inputs, sequence_id=i, sequence_start=True
+                            )
+                        except Exception as ex:
+                            self.assertTrue(
+                                False, "unexpected inference error {}".format(ex)
+                            )
+
+                    # Each instance should be busy until their sequence times out, so
+                    # an additional infer call should time out. If it doesn't time out, something
+                    # is wrong and the test should fail.
+                    callable = partial(
+                        triton_client.infer,
+                        model,
+                        inputs,
+                        sequence_id=num_instances + 1,
+                        sequence_start=True,
+                    )
+                    self._call_with_expected_timeout(callable, timeout_secs=3)
+
+                    # Unload the model
+                    try:
+                        triton_client.unload_model(model)
+                    except Exception as ex:
+                        self.assertTrue(False, "unexpected error {}".format(ex))
+
+                    # Allow server to fully unload model before next test iteration
+                    num_tries = 10
+                    for i in range(num_tries):
+                        if triton_client.is_server_ready():
```

Comment:
Does the server become not ready when a model is unloaded? I was wondering whether we should make `unload_model` blocking? I think it is really confusing that it is async. (cc @kthui @GuanLuo). Also, even if it is async we should probably update this example to better demonstrate the usage: https://github.com/triton-inference-server/client/blob/f7c45d382639f31a5162666034fae7852681f082/src/python/examples/simple_grpc_model_control.py#L98-L101

Reply:
> Does the server become not ready when a model is unloaded?
Server ready is true when all models are ready for inferencing, so while a model is in the unloading state, server is considered not ready.

> I was wondering whether we should make unload_model blocking? I think it is really confusing that it is async.

I have been thinking about this recently when seeing Jacky's model update tests requiring polling, Kris's BLS test unload issues, and this test as well. I second that it's definitely confusing and unintuitive. 

I do believe we document that it returns without waiting for model to finish in most places. However, I still think it is unituitive and not symmetric with `load_model` which is blocking.

Reply:
NOTE: I understand there's good reason for an async unload to allow inflight requests to finish without a long blocking/wait time, but it would be nice to have an alternative or optional blocking call to support both.

Reply:
My reasoning is that when `unload` returns, the model is no longer accessible from the client. Does it matter from the client's perspective that whether the model is fully cleaned up on server side when `unload` returns?